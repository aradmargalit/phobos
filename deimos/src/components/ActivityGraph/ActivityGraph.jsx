import './ActivityGraph.scss';

import { Spin } from 'antd';
import React, { useEffect, useState } from 'react';
import {
  CartesianGrid,
  Line, LineChart, Tooltip, XAxis, YAxis,
} from 'recharts';

import { fetchMonthlySums } from '../../apis/phobos-api';

export default function ActivityGraph() {
  const [monthlyData, setMonthlyData] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetchMonthlySums(setMonthlyData, setLoading);
  }, [setLoading]);

  const transform = (data) => data.map(({ month, duration }) => (
    { month, duration: parseFloat((duration / 60).toFixed(2)) }
  ));

  return (
    <Spin spinning={loading}>
      <LineChart
        className="activity-graph"
        width={1000}
        height={500}
        data={transform(monthlyData)}
        margin={{
          top: 5, right: 30, left: 20, bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="10 10" />
        <XAxis tick={false} dataKey="activity_date" />
        <YAxis />
        <Tooltip />
        <Line type="monotone" dataKey="duration" stroke="#de541e" dot={false} strokeWidth={3} />
      </LineChart>
    </Spin>

  );
}


/*
import React, { PureComponent } from 'react';
import {
  LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend,
} from 'recharts';

const data = [
  {
    name: 'Page A', uv: 4000, pv: 2400, amt: 2400,
  },
  {
    name: 'Page B', uv: 3000, pv: 1398, amt: 2210,
  },
  {
    name: 'Page C', uv: 2000, pv: 9800, amt: 2290,
  },
  {
    name: 'Page D', uv: 2780, pv: 3908, amt: 2000,
  },
  {
    name: 'Page E', uv: 1890, pv: 4800, amt: 2181,
  },
  {
    name: 'Page F', uv: 2390, pv: 3800, amt: 2500,
  },
  {
    name: 'Page G', uv: 3490, pv: 4300, amt: 2100,
  },
];

export default class Example extends PureComponent {
  static jsfiddleUrl = 'https://jsfiddle.net/alidingling/xqjtetw0/';

  render() {
    return (
      <LineChart
        width={500}
        height={300}
        data={data}
        margin={{
          top: 5, right: 30, left: 20, bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="name" />
        <YAxis />
        <Tooltip />
        <Legend />
        <Line type="monotone" dataKey="pv" stroke="#8884d8" activeDot={{ r: 8 }} />
        <Line type="monotone" dataKey="uv" stroke="#82ca9d" />
      </LineChart>
    );
  }
}
*/
