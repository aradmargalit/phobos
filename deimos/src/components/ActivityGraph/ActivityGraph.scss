.activity-graph {
  margin: 20px auto;
}