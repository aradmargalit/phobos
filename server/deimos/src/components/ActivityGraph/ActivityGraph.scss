.activity-graph {
  margin: 20px auto;
}

.graph-header {
  display: flex;
  justify-content: center;
  margin-top: 25px;

  & > h2 {
    margin: 0px 10px;
  }
}
