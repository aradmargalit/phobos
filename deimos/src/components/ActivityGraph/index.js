import ActivityGraph from './ActivityGraph';

export default ActivityGraph;
