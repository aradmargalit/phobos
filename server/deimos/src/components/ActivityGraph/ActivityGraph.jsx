import './ActivityGraph.scss';

import { Spin } from 'antd';
import { meanBy as _meanBy } from 'lodash';
import moment from 'moment';
import React from 'react';
import {
  Area,
  AreaChart,
  CartesianGrid,
  ReferenceLine,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';

import AngledGraphTick from '../AngledGraphTick';

const transform = data =>
  data
    // Todo, convert these to a moment-compliant format
    .sort((a, b) => moment(a.month) - moment(b.month))
    .map(({ month, duration }) => ({
      month,
      rawDuration: duration,
      duration: parseFloat((duration / 60).toFixed(2)),
    }));

const calculateAverage = data => _meanBy(data, 'rawDuration') / 60;
const calculateCurrentMonth = data =>
  (moment().daysInMonth() * data[data.length - 1].duration) /
  moment(new Date()).date();

export default function ActivityGraph({ loading, monthlyData }) {
  if (loading) return <Spin />;
  const data = transform(monthlyData);

  return (
    <div className="activity-graph-wrapper">
      <div className="graph-header">
        <h2>M O N T H L Y</h2>
        <h2>W O R K O U T</h2>
        <h2>H O U R S</h2>
      </div>

      <AreaChart
        className="activity-graph"
        width={1250}
        height={500}
        data={data}
        margin={{ top: 10, right: 30, left: 0, bottom: 0 }}
        padding={{ top: 10, right: 30, left: 30, bottom: 10 }}
      >
        <defs>
          <linearGradient id="durationColor" x1="0" y1="0" x2="0" y2="1">
            <stop offset="5%" stopColor="#117088" stopOpacity={0.6} />
            <stop offset="95%" stopColor="#117088" stopOpacity={0} />
          </linearGradient>
        </defs>
        <XAxis
          interval={3}
          dataKey="month"
          height={120}
          tick={<AngledGraphTick />}
        />
        <YAxis />
        <CartesianGrid strokeDasharray="3 3" />
        <ReferenceLine
          y={calculateAverage(data)}
          stroke="red"
          strokeDasharray="3 3"
          label={{
            position: 'top',
            value: 'Monthly Average',
          }}
        />
        <ReferenceLine
          y={calculateCurrentMonth(data)}
          stroke="blue"
          strokeDasharray="3 3"
          label={{
            position: 'top',
            value: "Current Month's Projection",
          }}
        />
        <Tooltip
          separator={null}
          formatter={value => [`${value} Hours`, '']}
          animationDuration={300}
        />
        <Area
          dataKey="duration"
          type="monotone"
          stroke="#0e5a6d"
          fillOpacity={1}
          fill="url(#durationColor)"
          label={({ x, y, stroke, value }) => (
            <text
              x={x}
              y={y}
              dy={-15}
              stroke={stroke}
              fontSize={12}
              textAnchor="middle"
              fill="gray"
            >
              {value}
            </text>
          )}
        />
      </AreaChart>
    </div>
  );
}
