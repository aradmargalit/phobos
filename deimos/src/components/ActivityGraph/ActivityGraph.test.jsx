import { shallow } from 'enzyme';
import React from 'react';

import ActivityGraph from './ActivityGraph';

const generateComponent = props => {
  const defaultProps = {
    monthlyData: [{ month: 'January 2020', duration: 123 }],
  };

  const mergedProps = { ...defaultProps, ...props };

  return shallow(<ActivityGraph {...mergedProps} />);
};

describe('<ActivityGraph />', () => {
  it('renders without crashing', () => {
    generateComponent();
  });

  describe('monthly average', () => {
    it('calculates the correct average with one month', () => {
      const component = generateComponent();
      expect(component.find('ReferenceLine').prop('y')).toEqual(123 / 60);
    });

    it('calculates the correct average with multiple months', () => {
      const component = generateComponent({
        monthlyData: [
          { month: 'January 2020', duration: 60 },
          { month: 'February 2020', duration: 180 },
        ],
      });
      expect(component.find('ReferenceLine').prop('y')).toEqual(2);
    });
  });
});
