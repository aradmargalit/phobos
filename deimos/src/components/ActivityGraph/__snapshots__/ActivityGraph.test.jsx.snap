// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivityGraph /> monthly average calculates the correct average with multiple months 1`] = `ShallowWrapper {}`;

exports[`<ActivityGraph /> monthly average calculates the correct average with one month 1`] = `ShallowWrapper {}`;
